source/common/network/address_impl.h

```diff
@@ -6,16 +6,51 @@
 #include <cstdint>
 #include <string>
 
+#include "envoy/common/exception.h"
 #include "envoy/common/platform.h"
 #include "envoy/network/address.h"
 #include "envoy/network/socket.h"
 
 #include "common/common/assert.h"
+#include "common/common/statusor.h"
 
 namespace Envoy {
 namespace Network {
 namespace Address {
 
+template <typename InstanceType, typename... Args>
+StatusOr<InstanceConstSharedPtr> createInstancePtr(Args&&... args) {
+  absl::Status status;
+  status = InstanceType::validateProtocolsupported();
+  if (!status.ok()) {
+    return status;
+  }
+  std::shared_ptr<InstanceType> instance(new InstanceType(std::forward<Args>(args)...));
+  if (!instance->initError().ok()) {
+    return instance->initError();
+  }
+  return instance;
+}
+
+template <typename InstanceType, typename... Args>
+InstanceConstSharedPtr createInstancePtrOrThrow(Args&&... args) {
+  StatusOr<InstanceConstSharedPtr> error_or_address =
```

Comment:
Can you add a ASSERT(Thread::MainThread::isMainThread()) here?

Reply:
Fail several tests after I add the assertion...

Reply:
I assume that the fix for that is to change the offending call-sites to use the non-throwing version, right?

Reply:
Envoy::Upstream::OriginalDstCluster::LoadBalancer::requestOverrideHost(), this actually happens in worker thread.

Reply:
Will fix them, just switch to the error status version.

Reply:
I just realize these function call constructor that never throws. just add the main assertion before throw statement instead of at the  beginning of template method. No need to fix anything.